Integration with 3Dconnexion input devices

pskowronskiTDx opened the pull request:
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

The contribution was created in whole or in part by me and I
have the right to submit it under the open source license
indicated in the file.

I understand and agree that this project and the contribution
are public and that a record of the contribution (including all
personal information I submit with it, including my sign-off) is
maintained indefinitely and may be redistributed consistent with
this project or the open source license(s) involved.

ghutchis commented on avogadro/menubuilder.h:
If it's useful API, just add it, not `#ifdef`

ghutchis commented on avogadro/icons/3dx_pivot.png:
This seems kind of small. Does this work okay on a high-resolution / Retina screen?

pskowronskiTDx replied:
Yes, this icon is meant to be so small.

ghutchis reviewed (changes requested):
These are mostly minor, since the main aspect is your code itself.

ghutchis commented:
@pskowronskiTDx - comments on the requested changes? I'd certainly love to merge these.

pskowronskiTDx commented on avogadro/tdxcontroller.h:
```diff
@@ -0,0 +1,157 @@
+  This source code is released under the 3-Clause BSD License, (see "LICENSE").
+******************************************************************************/
+
+#pragma once
```
Of course, corrected.

ghutchis commented on thirdparty/3DConnexion/README.txt:
```diff
@@ -0,0 +1,3 @@
+This directory contains part of the the 3Dconnexion 3DxWare_SDK v4
+The full sdk can be obtained from https://3dconnexion.com/software-developer-program/
```
Is there a good way to keep this up-to-date (e.g., when v5 comes out?)

pskowronskiTDx replied:
No, as we are shipping some SDK's headers with the solution they will need to be updated manually when a new release comes out.

ghutchis commented on avogadro/tdxcontroller.cpp:
```diff
@@ -0,0 +1,592 @@
+
+  std::vector<bool> flags;
+
+  m_pGLRenderer->scene().getBoundingBox(extents.min.x, 
```
See, I think you can get this from the molecule here and avoid having to save all those spheres in the render pipeline.

Just implement `core/molecule.*:getBoundingBox()` instead.

pskowronskiTDx replied:
Done as suggested.

ghutchis commented on avogadro/tdxcontroller.cpp:
```diff
@@ -0,0 +1,592 @@
+  for (uint32_t i = 0u; i < (*m_ppMolecule)->atomCount(); i++)
+    flags[i] = (*m_ppMolecule)->atomSelected(i);
+
+  m_pGLRenderer->scene().getBoundingBox(extents.min.x,
```
Ditto. You're already going through the atoms in the loop above.

```
for (Index i = 0; i < molecule->atomCount(); ++i) {
   if (molecule->atomSelected(i) {
     // calculate the bounding box directly
   }
}
```

ghutchis commented on avogadro/tdxcontroller.cpp:
```diff
@@ -0,0 +1,592 @@
+
+long Avogadro::TDxController::GetIsSelectionEmpty(navlib::bool_t &empty) const
+{
+  if (*m_ppMolecule == nullptr)
```
Use `Molecule:isSelectionEmpty()`

ghutchis commented on avogadro/tdxcontroller.h:
```diff
@@ -0,0 +1,162 @@
+public:
+  TDxController(MainWindow* const mainWindow,
+                Avogadro::QtOpenGL::GLWidget* const pGLWidget,
+                QtGui::Molecule** ppMolecule);
```
I don't understand the double pointer. Is there a good reason for this?

pskowronskiTDx replied:
The double pointer is used to keep track of the actual pointer to active molecule. Then there is no need to update the pointer within the navigation object  each time the active molecule is switched in the explorer. Is there a different approach I should folllow?

ghutchis commented:
Looks pretty close - sorry for the delay. If you could run clang-format on your changed files in `avogadro` (i.e. outside of `thirdparty`) it would be appreciated.

pskowronskiTDx commented:
Thank you for your sugesstions, I have introduced some changes. The double pointer for Molecule object was used to keep track on the active molecule (selected in the explorer). However, I have reworked it by introducing updateMolecule method, connected to the moleculeChanged signal. The pointer to the active molecule updates automatically now, without referencing "pp" anymore.

ghutchis commented:
Assuming this builds cleanly, I'll merge this as well. Thanks!

ghutchis commented:
Sorry for the delay - I thought I had previously merged this before leaving for a conference. I'll update the clang-format myself in a new pull.

Thanks!

pskowronskiTDx commented:
No problem, I am happy to see the changes finally merged. Hope the integration will be useful. Thank you as well!
